os/bluestore: sync commit transaction

liewegas opened the pull request.

varadakari commented on src/os/bluestore/BlueStore.cc:
```diff
@@ -6230,29 +6236,40 @@ void BlueStore::_txc_state_proc(TransContext *txc)
-	  txc->kv_submitted = true;
+	  txc->state = TransContext::STATE_KV_SUBMITTED;
+	  if (g_conf->bluestore_rtc &&
+	      txc->osr->kv_finisher_submitting == 1) {
```
couldn't find kv_finisher_submitting. Need to add to the sequencer. Is it a bool?

varadakari replied:
@voidbag, i couldn't find the symbol in the PR, so got lost where it is getting modified. Thanks for the explanation. Will look at the other PR also.

varadakari commented:
This one has #11983 changes also.

voidbag commented:
@liewegas 
Could you let me know any plan for applying this patch?

Thanks

liewegas commented:
Yes, it's on hold for a week or two while we focus on getting kraken stable and released, then I'll come back to it.  Thanks for your patience!

voidbag commented:
Thanks for your detailed and clear explanation!

voidbag commented:
@liewegas
ping

liewegas commented:
@voidbag rebase and updated.  I've added an option to push small writes through WAL even when we don't strictly have to so that the sync commit path is actually taken.  This seems to work for me.  The main remaining issue is that the completions are not yet synchronous.

voidbag commented:
@liewegas
Synchronous commit looks good to me!
I'll check the code about synchronous completion

Thanks!

voidbag commented:
I wonder if we don't have to handle below cases.
One(FlushState) takes a PG lock in destructor and the other(RunOnDelete) calls complete callbacks in destructor.

src/osd/PG.cc
![image](https://cloud.githubusercontent.com/assets/6831102/21509641/442c64a2-ccce-11e6-9864-87bb29f477d0.png)

src/include/Context.h
![image](https://cloud.githubusercontent.com/assets/6831102/21509657/6ae9f384-ccce-11e6-8780-d7ae3fee2a63.png)

I experienced assert failures, when i implemented sync_completions with passing FLAG_SYNC
I resolved them by applying sync completions only to oncommit, not onreadable.
That's because there is no problem, when finisher calls above destructors which try to take PG lock.

So, I wonder if we don't have to consider above cases...

Thanks

markhpc commented:
retest this please

markhpc commented:
This no longer applies cleanly.  We now check: !txc->oncommits.empty() and pass in txc->oncommits instead of looping over txc->oncommits and passing in each item.  The problem is that this code wants to add some conditional checks inside that loop but we got rid of it.  Might be some other problems as well but that's the first one I found.

liewegas commented:
I'll rebase

markhpc commented:
Latest code built for me, but failed quickly after OSD creation:

` -4276> 2017-01-18 14:06:02.189834 7f8f6f6b4700 -1 *** Caught signal (Aborted) **
 in thread 7f8f6f6b4700 thread_name:finisher

 ceph version 10.0.4-18972-g2bf82c5 (2bf82c5c07d79d5e78d369d34a5639032b181ee5)
 1: (()+0x7c3ccf) [0x556a3bd48ccf]
 2: (()+0xf100) [0x7f8f81073100]
 3: (gsignal()+0x37) [0x7f8f7fc7b5f7]
 4: (abort()+0x148) [0x7f8f7fc7cce8]
 5: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x7f8f805809d5]
 6: (()+0x5e946) [0x7f8f8057e946]
 7: (()+0x5e973) [0x7f8f8057e973]
 8: (()+0x5f4df) [0x7f8f8057f4df]
 9: (Context::complete(int)+0x9) [0x556a3b8b0209]
 10: (Finisher::finisher_thread_entry()+0x1d0) [0x7f8f825262c0]
 11: (()+0x7dc5) [0x7f8f8106bdc5]
 12: (clone()+0x6d) [0x7f8f7fd3cced]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
`

bluestore debug 20 log is on incerta01 in ~/wip-rtc

voidbag commented:
@liewegas
I think the problem might be resolved by using sync completions only for oncommit callback, not for onreadable callback, as I implemented.
Callback in destructor had to be handled by finisher to avoid taking duplicate pg lock.

rchagam commented:
@liewegas tested performance improvement with bluestore_sync_submit_transaction = true and bluestore_sync_commit_transaction = true. 
Code below in _txc_state_proc does not get executed at all
 else {
          _txc_finalize_kv(txc, txc->t);
          txc->state = TransContext::STATE_KV_SUBMITTED;
          if (g_conf->bluestore_sync_commit_transaction &&
              txc->osr->txc_with_unsubmitted_completions.load() == 1) {
            db->submit_transaction_sync(txc->t);
            _txc_release_alloc(txc);
            _txc_state_proc(txc);
            return;
          }
          int r = db->submit_transaction(txc->t);
          assert(r == 0);

**Either else if (txc->osr->kv_committing_serially) or else if (txc->osr->txc_with_unstable_io) { gets executed. Looking at the logic txc->had_ios = true; is always going to be true in case TransContext::STATE_IO_DONE:

One way to fix this is reset had_ios in 
case TransContext::STATE_IO_DONE:
**if (!txc->ioc.has_pending_aios())
        txc->had_ios = false;**

rchagam commented:
@liewegas bluestore_sync_commit_transaction = true causes rados client to hang and can't get below code to work properly 
void BlueStore::_txc_state_proc(TransContext *txc)
{
..
**if (g_conf->bluestore_sync_commit_transaction &&
              txc->osr->txc_with_unsubmitted_completions.load() == 1) {
            db->submit_transaction_sync(txc->t);
            _txc_release_alloc(txc);
            _txc_state_proc(txc);
            return;
          }
**
..
}

rchagam commented:
Once above two start working, we will see aio_wait callback latency will significantly increase considering it blocks _aio_thread() aio_callback(..) will block for completion of kv sync transaction. This is expected. 
One option is to have separate thread pool for sync transactions and it will likely be quite a number of threads if it is NVMe SSD device to keep up the IO rate. Other option is to consolidate threading for both sync and async transaction execution (_kv_sync_thread() changes). 
Ideally if we can get away with one thread pool for sync transactions and bypass finisher pool by calling op_applied or op_commit callbacks in OSD this will cut down latency drastically.

liewegas commented:
Did you have bluestore_prefer_wal_size set (to 32k or so)?

liewegas commented:
What was the workload?

voidbag commented:
@rchagam 
The optimization is for update workloads (WAL case).

Don't  modify code as  txc->had_ios = false;
I ask u to let txc->had_ios = true.

configuration
bluestore_min_alloc_size = 4194304 (=4M)
You can execute 4k random write fios, after u fill the whole rbd image.

If you don't want to fill the whole rbd image, you can set bluestore_prefer_wal_size 32K as @liewegas  mentioned.
Then,
            db->submit_transaction_sync(txc->t);
            _txc_release_alloc(txc);
            _txc_state_proc(txc);
            return;
will be executed.

I already implemented the option you(@rchagam) mentioned and tested it with performance result.
I have a plan for pushing a new PR about it.

Thanks

rchagam commented:
@liewegas I used rados bench 4K writes (so no updates) and default bluestore_prefer_wal_size. Let me try with fio random writes and set wal_size to 32K

voidbag commented:
@liewegas 
I ask you to separate sync_commit and sync_completions as i implemented, not merge the two options 
I think we have to make tuning parameters fine-grained.

Thanks

voidbag commented:
@liewegas 

I left a message to remind you of the patches.
Would you let me know the plan for adapting these patches (rtc & sync_completions)?
I want to know the progress about patches.

Thank you for reading it.

varadakari commented:
needs a rebase

liewegas commented:
Hi @voidbag sorry, I've been sitting on this for a while because i can't figure out if this is going to have an impact for real workloads and because i'm worried about adding complexity here if it's going to pay off, and there been plenty of other stuff to keep me distracted.  Sorry for the delay!

It would help if you could tell/remind me a bit about what hardware you're using where you see this helps. Is it a theoretical config (actual RAM for wal) or an actual NVMe?  Which one?  What workload?

Thanks!

voidbag commented:
@liewegas 
Thank you for replying, despite being busy.
I'm using sata3 SSD(850 pro) as block device.
Ramdisk is used as db and wal devices. Ramdisk is considered as NVDIMM.
It's not theoretical config.

[config]
bluestore_kvbackend = rocksdb
bluestore_bluefs = true
bluestore_min_alloc_size = 4194304
bluestore_sync_submit_transaction = true
bluestore_csum_type = none
bluefs_buffered_io = true
bluestore_sync_commit_transaction = true
bluestore_sync_wal_apply = true
**bluestore_sync_completions = [true or false].(in case of me, false is better)**
other throttling options are big enough to allow all requests.
Other options are default.

[Hardware]
CPU: i7-4790
Block Device: Samsung 850 Pro (256G or 512G) (Sata3 SSD, not NVMe)
DB and Wal Device: Ramdisk 4G for each other (2G is also okay)
One OSD per Node.
2 10G NICs per Node (one for cluster, the other for public)

[Workload]
4K random (128 threads) direct write(fio) on rbd, after 4M sequential write with 64 IO-depth(fio) on the whole rbd.
Above workload is "update" for using only WAL path. 4M sequential write may be unnecessary if bluestore_prefer_wal_size is used.

[Performance Result]
fio 128 threads, synchronous, direct 4K random I/O
- 1 OSD, 1 RBD Client
Maximum Throughput: 52MB/s -> 71MB/s (13KIOPS -> 17.75 KIOPS)
![image](https://cloud.githubusercontent.com/assets/6831102/23385780/909e492a-fd95-11e6-9ddd-ce7a49eb0c07.png)


- 2 OSDs, 2 Replicas, 1 RBD Client
Maximum Throughput: 46MB/s -> 60MB/s (11.5KIOPS -> 15 KIOPS)
![image](https://cloud.githubusercontent.com/assets/6831102/23385782/9879bb02-fd95-11e6-9f01-7b1544dd907e.png)


=================================================================

This patch is good for **commodity SSD without supercapacitor**. If **NVMe SSD with supercapacitor** is used as block device, it has little impact on performance and adds unnecessary complexity. That's because there is no flush command overhead, when **NVMe SSD with supercapacitor** is used.

The key of this patch is to remove bdev->flush() in the critical path of WAL write. The overhead bdev->flush() of SSD w/o supercapacitor(850 Pro) is huge. **So, this patch improves performance (4K random write - WAL path)** 

Note:
There is little benefits in case of SSD w/ supercapacitor, which has **no flush command overhead.**

Thanks :)

liewegas commented:
@voidbag The "Ramdisk is used as db and wal devices. Ramdisk is considered as NVDIMM.
It's not theoretical config." part is what troubles me; if you're using volatile RAM that seems theoretical to me!  Or are you simulating the case of battery-backed NVDIMM or persistent memory here?

Either way, the performance gains here are pretty significant.  This is next up on my list after I finish fixing the deferred write ordering and flush issues.  https://github.com/ceph/ceph/pull/13888

voidbag commented:
@liewegas 
Sorry for confusing.
I just used just DRAM for ramdisk, not NVDIMM. And I didn't simulate anything.
Unlike PCM, the latency of NVDIMM is similar to DRAM.
http://www.snia.org/sites/default/files/SSSI/NVDIMM%20-%20Changes%20are%20Here%20So%20What's%20Next%20-%20final.pdf

That's why i considered DRAM Ramdisk as NVDIMM Ramdisk...

Thanks

liewegas commented:
Okay, so the target is NVDIMM, not pmem.

Putting the wal in DRAM is totally unsafe and will corrupt the store if you lose power.  If you're not using an actual NVDIMM, you're "simulating" what someone who went and spent the $$ would be able to do.

Which is all fine.. we should just be clear that the performance benefits here are only available to users in the real world if they use NVDIMMs.

voidbag commented:
That's right!
The performance gain is for the user who can buy and use NVDIMM.

Thanks

voidbag commented:
@liewegas 
Hi, Sage.
I checked finalizing thread in BlueStore has been added.
Any plan for this PR?

Thanks.

liewegas commented:
I will rebase.  I think the first part of this series might make sense for luminous but the finisher pieces will have to wait

liewegas commented:
rebased, ready to test

liewegas commented:
retest this please

tchaikov commented:
retest this please

tchaikov commented:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/standalone/ceph-helpers.sh:1263: wait_osd_id_used_by_pgs:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/ceph-disk/tests/ceph-disk.sh:318: test_reuse_osd_id:  return 1
```
